libnopegl/src/node_hblur.c

```diff
@@ -0,0 +1,677 @@
+    ngli_assert(dst_priv->params.format == s->pass2.layout.colors[0].format);
+
+    dst = dst_priv->texture;
+    if (s->dst_is_resizeable) {
+        dst = ngli_texture_create(ctx->gpu_ctx);
+        if (!dst) {
+            ret = NGL_ERROR_MEMORY;
+            goto fail;
+        }
+
+        struct texture_params params = dst_priv->params;
+        params.width = width;
+        params.height = height;
+        ret = ngli_texture_init(dst, &params);
+        if (ret < 0)
+            goto fail;
+    }
+
+    if (s->dst_is_resizeable) {
+        ngli_texture_freep(&dst_priv->texture);
+        dst_priv->texture = dst;
+        dst_priv->image.params.width = dst->params.width;
+        dst_priv->image.params.height = dst->params.height;
+        dst_priv->image.planes[0] = dst;
+        dst_priv->image.rev = dst_priv->image_rev++;
+    }
+
+    pass2_rtt_ctx = ngli_rtt_create(ctx);
+    if (!pass2_rtt_ctx) {
+        ret = NGL_ERROR_MEMORY;
+        goto fail;
+    }
+
+    const struct rtt_params pass2_rtt_params = (struct rtt_params) {
+        .width  = dst->params.width,
+        .height = dst->params.height,
+        .nb_colors = 1,
+        .colors[0] = {
+            .attachment = dst,
+            .load_op = NGLI_LOAD_OP_CLEAR,
+            .store_op = NGLI_STORE_OP_STORE,
+        },
+    };
+
+    ret = ngli_rtt_init(pass2_rtt_ctx, &pass2_rtt_params);
+    if (ret < 0)
+        goto fail;
+
+    ngli_rtt_freep(&s->pass2.rtt_ctx);
+    s->pass2.rtt_ctx = pass2_rtt_ctx;
+
+    ngli_rtt_freep(&s->pass1.rtt_ctx);
+    s->pass1.rtt_ctx = pass1_rtt_ctx;
+
+    ngli_texture_freep(&s->tex0);
+    s->tex0 = tex0;
+
+    ngli_texture_freep(&s->tex1);
+    s->tex1 = tex1;
+
+    ngli_pipeline_compat_update_texture(s->pass2.pl, 0, s->tex0);
+    ngli_pipeline_compat_update_texture(s->pass2.pl, 1, s->tex1);
+
+    s->width = width;
+    s->height = height;
+
+    return 0;
+
+fail:
+    ngli_rtt_freep(&pass1_rtt_ctx);
+    ngli_texture_freep(&tex0);
+    ngli_texture_freep(&tex1);
+
+    ngli_rtt_freep(&pass2_rtt_ctx);
+    if (s->dst_is_resizeable)
+        ngli_texture_freep(&dst);
+
+    LOG(ERROR, "failed to resize blur: %dx%d", width, height);
+    return ret;
+}
+
+#define MAX_SAMPLES 32
+
+static void hblur_draw(struct ngl_node *node)
+{
+    struct ngl_ctx *ctx = node->ctx;
+    struct gpu_ctx *gpu_ctx = ctx->gpu_ctx;
+    struct hblur_priv *s = node->priv_data;
+    struct hblur_opts *o = node->opts;
+
+    int ret = resize(node);
+    if (ret < 0)
+        return;
+
+    const float amount_raw = *(float *)ngli_node_get_data_ptr(o->amount_node, &o->amount);
+    float amount = NGLI_CLAMP(amount_raw, 0.f, 1.f);
+    const float diagonal = hypotf((float)s->width, (float)s->height);
+    const int32_t radius = (int32_t)(amount * (float)(diagonal) * 0.05f);
+    const int32_t nb_samples = NGLI_MIN(radius, MAX_SAMPLES);
+
+    ngli_gpu_block_update(&s->blur_params_block, 0, &(struct blur_params_block) {
+        .radius = (int)radius,
```

Comment:
I don't think the cast is necessary, both field and local variables are `i32`

Reply:
Removed.